[DOCS-3273] GA for monitor RBAC (#13446)
DataDog/documentation: content/en/monitors/guide/how-to-set-up-rbac-for-monitors.md

---
title: How to set up RBAC for Monitors
kind: guide
further_reading:
- link: "/account_management/rbac/permissions/#monitors"
  tag: "Documentation"
  text: "Learn more about RBAC permissions for Monitors"
- link: "/api/latest/monitors/#create-a-monitor"
  tag: "Documentation"
  text: "Learn more about creating restricted monitors via API"
- link: "/monitors/notify/#permissions"
  tag: "Documentation"
  text: "Learn more about creating restricted monitors via the UI"
---

## Overview

Monitors ensure your teams get alerted in case of issues with your systems. Making sure only the right subset of users can edit the right subset of monitors is critical to **avoid accidental changes in your monitors' configurations**.

> Safely manage your monitors by **restricting edit permissions of each individual monitor to some specific roles** of your account.

Author: Why keeping the bold in the first sentence but not that one?

> Safely manage your monitors by **restricting edit permissions of each individual monitor to some specific roles** of your account.

Reviewer: I meant to take it out of both, but I don't feel strongly about it. You can leave the bold in.

{{% /tab %}}

{{% tab "API" %}}

Use the [List Roles API endpoint][1] to get the list of roles and their id.

```bash
curl --request GET 'https://api.datadoghq.com/api/v2/roles' \
--header 'DD-API-KEY: <DD-API-KEY>' \
--header 'DD-APPLICATION-KEY: <DD-APPLICATION-KEY>'
```

```bash
{
    "meta": {
        "page": {
            "total_filtered_count": 4,
            "total_count": 4
        }
    },
    "data": [
        {
            "type": "roles",
            "id": "89f5dh86-e470-11f8-e26f-4h656a27d9cc",
            "attributes": {
                "name": "Corp IT Eng - User Onboarding",
                "created_at": "2018-11-05T21:19:54.105604+00:00",
                "modified_at": "2018-11-05T21:19:54.105604+00:00",
                "user_count": 4
            },
            "relationships": {
                "permissions": {
                    "data": [
                        {
                            "type": "permissions",
                            "id": "984d2rt4-d5b4-13e8-a5yf-a7f560d33029"
                        },
                        ...
                    ]
                }
            }
        },
        ...
    ]
}
```

> Use the [Create a monitor][2] or [Edit a monitor][3] API endpoint and the `restricted_roles` parameter to restrict the edition of your monitor to a specific set of roles and to the monitor's creator. 

Author: ```suggestion
Use the [Create][2] or [Edit a monitor][3] API endpoint and the `restricted_roles` parameter to restrict monitor edition to a specific set of roles and to the monitor's creator. 
```

**Note:** You can specify one or multiple role UUIDs. Setting `restricted_roles` to `null` allows monitor edition for all users with [Monitor Write permissions][4].

```bash
curl --location --request POST 'https://api.datadoghq.com/api/v1/monitor' \
--header 'Content-Type: application/json' \
--header 'DD-API-KEY: <DD-API-KEY>' \
--header 'DD-APPLICATION-KEY: <DD-APPLICATION-KEY>' \
--data-raw '{
  "message": "You may need to add web hosts if this is consistently high.",
  "name": "Bytes received on host0",
  "options": {
    "no_data_timeframe": 20,
    "notify_no_data": true
  },
  "query": "avg(last_5m):sum:system.net.bytes_rcvd{host:host0} \u003e 100",
  "tags": [
    "app:webserver",
    "frontend"
  ],
  "type": "query alert",
  "restricted_roles": ["89f5dh86-e470-11f8-e26f-4h656a27d9cc"]
}'
```

For more information, see [Roles][5] and [Monitors API documentation][6] .

[1]: /api/latest/roles/#list-roles
[2]: /api/latest/monitors/#create-a-monitor
[3]: /api/latest/monitors/#edit-a-monitor
[4]: /account_management/rbac/permissions/#monitors
[5]: /api/latest/roles/
[6]: /api/latest/monitors/
{{% /tab %}}
{{< /tabs >}}

> ## Deprecated locked mechanism

Reviewer: ```suggestion
## Migrate locked monitors to use RBAC
```

> ## Migrate locked monitors to use RBAC

Author: Actually that title does not really work since we also cover non locked monitors in the UI section.

Note to self before closing this: make sure to update any link that was referring to this.

> Previous to the above described role restriction option, monitors could be locked, meaning only their creator and users with the [Datadog Admin Role][2] would be able to edit them. 

Reviewer: ```suggestion
Before Datadog released the feature allowing restriction of monitor editing to specific roles, monitors could be locked. Only the creator and users with the [Datadog Admin Role][2] can edit a locked monitor. 
```

{{< img src="/monitors/guide/monitor_rbac_locked.jpg" alt="RBAC Locked Monitor" style="width:70%;">}}

> This locked mechanism was deprecated and it is now recommended to use the role restriction option, which gives you more flexibility to define users allowed to edit monitors.

Reviewer: ```suggestion
Locked monitors are deprecated. Datadog recommends using the role restriction option, which gives you more flexibility to define the users allowed to edit monitors.
```

> You might however have a number of existing monitors still leveraging the locked mechanism under the hood. This mechanism is still supported and respected for now, which means locked monitors should not be editable by users who do not have the [Datadog Admin Role][2] attached or are not the monitor's creator.

Reviewer: ```suggestion
Your organization may have existing locked monitors. Datadog still supports locked monitors. Editing of locked monitors is restricted to users with the [Datadog Admin Role][2] and the monitor's creator.
```

> Below sections describe how the change from the old locked mechanism to the new role restriction one is performed, depending on the way you manage your monitors.

Reviewer: ```suggestion
The sections below describe how to migrate from the locked mechanism to restricted roles, depending on the way you manage your monitors.
```

### API

> Although deprecated, the `locked` parameter corresponding to the above mentioned locked mechanism is still supported. This means you can progressively update the definition of your monitors managed through API/Terraform to stop using `locked` and start using `restricted_roles` (parameter attached with the new role restriction option).

Reviewer: ```suggestion
Although it is deprecated, the `locked` parameter is still supported. This means you can progressively update the definitions of your monitors managed through the API or Terraform to stop using `locked` and start using `restricted_roles`.
```

Author: I wanted to introduce the fact that we referred to as the role restriction option and the locked one corresponded to `restricted_roles` and `locked` API parameters.
Do you think that's overkill? Or should we add that as an intro sentence?

> Although deprecated, the `locked` parameter corresponding to the above mentioned locked mechanism is still supported. This means you can progressively update the definition of your monitors managed via API/Terraform to stop using `locked` and start using `restricted_roles` (parameter attached with the new role restriction option).

Author: ```suggestion
Although deprecated, the `locked` parameter corresponding to the above mentioned locked mechanism is still supported. This means you can progressively update the definition of your monitors managed through API/Terraform to stop using `locked` and start using `restricted_roles` (parameter attached with the new role restriction option).
```

> For more information on how to update your monitors' definitions, see the [Edit a monitor API endpoint documentation][3] and the dedicated [Monitor API Options guide][4].

Reviewer: ```suggestion
For more information on how to update your monitors' definitions, see the [Edit a monitor API endpoint][3] and the [Monitor API Options guide][4].
```

Author: > see the Edit a monitor API endpoint

Doesn't that sound weird? What about "see Edit a monitor API endpoint"? (It might not sound weird for an English speaker though! It felt weird to me so wanted to flag but let me know if it is actually ok!)

> Monitor API Options guide

Based on your other comments, it looks like we are trying to move away from saying "see XYZ **documentation**", why is it ok to mention guide here?

Reviewer: ```suggestion
For more information on how to update your monitors' definitions, see [Edit a monitor API endpoint][3] and the [Monitor API Options][4].
```
"Guide" is more specific than "documentation". Sorry, it's hard to remember to be consistent across a whole review. But you're right, to be consistent with the style guide, it's better to remove "Guide" also.

I left in the "the" by accident.

How about the phrasing above?

Author: LGTM, committing!

### UI

> All new monitors created from the UI use the new `restricted_roles` parameter. 

Reviewer: ```suggestion
All new monitors created from the UI use the `restricted_roles` parameter. 
```

> All monitors (existing and new) also display the new role restriction option regardless of the mechanism used under the hood:

Reviewer: ```suggestion
All monitors also display the role restriction option regardless of the underlying mechanism:
```

{{< img src="/monitors/guide/monitor_rbac_non_restricted.jpg" alt="RBAC Non Restricted Monitor"  >}}

> Existing monitors have their definitions updated from the old locked mechanism to the new role restriction one on the fly whenever a monitor is saved.

Reviewer: ```suggestion
Datadog updates existing monitor definitions from the old locked mechanism to the new role restriction one whenever a monitor is saved.
```

> Below are some instructions to help you determine how to proceed in case you need to save a monitor that is using the previous locked mechanism.

Reviewer: ```suggestion
Below are some instructions to help you determine how to proceed in case you need to save a monitor that is using the locked mechanism.
```

#### Locked monitors (`locked:true`) edited by creator or user with Datadog Admin Role

You are a user with [Datadog Admin Role][2] and/or are the creator of the monitor. You are editing a locked monitor and seeing the below warning:

> <TO DO: ADD SCREENSHOT OF WARNING>

Author: ```suggestion
This monitor is using the `locked` attribute: only its creator and admins can edit it. `locked` is deprecated in favor of `restricted_roles`. On save, the monitor will be automatically updated to use a `restricted_roles` attribute set to all roles with Admin permissions. 
If there is no specific change you want to apply to this monitor’s permissions, click Save. If you want to update this monitor’s permissions, read [this doc](https://docs.datadoghq.com/monitors/guide/how-to-set-up-rbac-for-monitors/?tab=ui#locked-monitors-lockedtrue-edited-by-creator-or-user-with-datadog-admin-role).
```

> You are a user with [Datadog Admin Role][2] and/or are the creator of the monitor. You are editing a locked monitor and seeing the below warning:

Reviewer: ```suggestion
You are a user with [Datadog Admin Role][2] and/or are the creator of the monitor. You edit a locked monitor and seeing the below warning:
```

Reviewer: oops, yes, it should be "and see"

Author: Coolio, committing!

> <TO DO: ADD SCREENSHOT OF WARNING>

Reviewer: Please remove this TODO and replace it with the text of the warning. Whenever possible, the Docs team favors text descriptions over screenshots. Text descriptions are more accessible to low-vision folks and those using screen readers, and we don't need to update them as often. The UI changes often, making screenshots outdated.

Author: Done

On save, your monitor's definition will be updated to all roles with Admin permissions.

> You have different ways to handle this warning depending on the changes you are willing to make on your monitor:

Reviewer: ```suggestion
You have different ways to handle this warning depending on the changes you are willing to make to your monitor:
```

**1. You do not want to change anything about your monitor permissions**

> Save the monitor to have the update from the locked mechanism to the restricted_roles one be performed automatically, along with any other updates you might have done on your monitor (e.g. threshold update, message, etc.).

Reviewer: ```suggestion
Save the monitor. Datadog automatically migrates the monitor from the locked mechanism to restricted roles. Any other updates you made to your monitor, such as threshold update or message, are saved at the same time.
```

> If you were just opening the edit page to see more details about your monitor configuration, you can also hit **Save**. That also performs the update for your monitor.

Reviewer: ```suggestion
Clicking **Save** without making any changes also performs the update for your monitor.
```

**2. You want to allow all users to edit this monitor**

> Save the monitor, edit it again, and remove all roles that were generated for this monitor. 

Reviewer: ```suggestion
Save the monitor, causing Datadog to migrate it to restricted roles. Reopen the editing page. In the **Restrict editing of this monitor to** dropdown menu, remove all roles. Click **Save** a second time.
```

> **3. You want your monitor to be restricted to some roles, but not the exact equivalent of the previous locked option**

Reviewer: ```suggestion
**3. You want to change the roles allowed to modify your monitor**
```

Author: This makes it sound like the monitor is already using the restricted role mechanism at this point, which it's not.

> Select the roles you want this monitor to be restricted to. This will override the locked equivalent set of roles to only the ones you specify.

Reviewer: ```suggestion
In the **Restrict editing of this monitor to** dropdown menu, select the roles that can modify this monitor. Datadog replaces the generated set of roles with the ones you specify.
```

  
#### Locked monitors (`locked:true`) edited by non creator or user without Datadog Admin Role

You are a user without [Datadog Admin Role][2] and/or are not the creator of the monitor. You are editing a locked monitor and seeing the below warning:

> <TO DO: ADD SCREENSHOT OF WARNING>

Author: Done

> As mentioned in the warning, this monitor was locked in the previous system. You should consequently reach out to a user with a [Datadog Admin Role][2] or to the creator of the monitor for them to update the monitor to include one of the roles you belong to. Your admin will have to follow guidelines in the [above section](#locked-monitors-lockedtrue-edited-by-creator-or-user-with-datadog-admin-role) (2. or 3.).

Reviewer: ```suggestion
This monitor is locked. Reach out to a user with the [Datadog Admin Role][2] or to the creator of the monitor and ask them to add one of your roles to the monitor role restrictions. Your admin will have to follow steps two or three above for [locked monitors](#locked-monitors-lockedtrue-edited-by-creator-or-user-with-datadog-admin-role).
```

> **Note:** The discrepancy you are seeing between the warning and the option is expected:

Reviewer: ```suggestion
**Note:** The discrepancy you see between the warning and the option is expected:
```

Author: Done

> <TO DO: ADD SCREENSHOT OF WARNING AND OPTION>

Reviewer: Replace the TO DO with the text of the warning.

Author: Completely got rid of that one, I don't think it's necessary to add the warning as it's mentioned above.

> The warning reflects the current state of the monitor that is still using the old locked parameter. The option reflects the new role restriction option your monitor will be updated to once a user with a [Datadog Admin Role][2]/the monitor's creator edits and saves it. Once saved, the warning will disappear and the adequate restricted roles will populate the dropdown.

Reviewer: ```suggestion
The warning reflects the current state of the monitor that is using the old locked parameter. The option reflects the new role restriction option your monitor will be updated to once a user with a [Datadog Admin Role][2] or the monitor's creator edits and saves it. Once the monitor is saved, the warning disappears and the appropriate restricted roles populates the dropdown.
```

Author: Done

#### Non locked monitors (`locked:false`, `locked:null`, undefined `locked`)

> You are editing a non locked monitor and seeing the below option:

Reviewer: ```suggestion
While editing a non locked monitor, you see the option below:
```

Author: Adjusted your suggestion a little to map wording used in above section

{{< img src="/monitors/guide/monitor_rbac_non_restricted.jpg" alt="RBAC Non Restricted Monitor"  >}}

You have different ways to handle this option depending on the changes you are willing to make on your monitor:

**1. You do not want to change anything about your monitor permissions**

> Save the monitor to have the update from the locked mechanism to the restricted_roles one be performed automatically, along with any other updates you might have done on your monitor (e.g. threshold update, message, etc.).
> 
> If you were just opening the edit page to see more details about your monitor configuration, you can also hit **Save**. That also performs the update for your monitor.

Reviewer: ```suggestion
Save the monitor. Datadog automatically migrates the monitor from the locked mechanism to restricted roles. Any other updates you made to your monitor, such as threshold update or message, are saved at the same time.

Clicking **Save** without making any changes also performs the update for your monitor.
```

**2. You want to restrict your monitor to some roles**

> Select the roles you want this monitor to be restricted to and save your monitor. The update of your monitor from non locked to restricted to some roles is performed automatically on save.

Reviewer: ```suggestion
In the **Restrict editing of this monitor to** dropdown menu, select the roles that can modify this monitor. Datadog automatically migrates the monitor from the locked mechanism to restricted roles when you click **Save**.
```

Author: Feels like we lose a bit the fact that the monitor will be using restricted roles mechanism + will be set to the selected roles. WDYT?

Reviewer: Yes, that's a good point. Please reject any of my suggestions that don't work or change the meaning. You know this material much better than I do.

**Note:** You can specify one or multiple role UUIDs. Setting `restricted_roles` to `null` allows monitor editing for all users with [Monitor Write permissions][4].

```bash
curl --location --request POST 'https://api.datadoghq.com/api/v1/monitor' \
--header 'Content-Type: application/json' \
--header 'DD-API-KEY: <DD-API-KEY>' \
--header 'DD-APPLICATION-KEY: <DD-APPLICATION-KEY>' \
--data-raw '{
  "message": "You may need to add web hosts if this is consistently high.",
  "name": "Bytes received on host0",
  "options": {
    "no_data_timeframe": 20,
    "notify_no_data": true
  },
  "query": "avg(last_5m):sum:system.net.bytes_rcvd{host:host0} \u003e 100",
  "tags": [
    "app:webserver",
    "frontend"
  ],
  "type": "query alert",
  "restricted_roles": ["89f5dh86-e470-11f8-e26f-4h656a27d9cc"]
}'
```

For more information, see [Roles][5] and [Monitors API Reference][6] .

[1]: /api/latest/roles/#list-roles
[2]: /api/latest/monitors/#create-a-monitor
[3]: /api/latest/monitors/#edit-a-monitor
[4]: /account_management/rbac/permissions/#monitors
[5]: /api/latest/roles/
[6]: /api/latest/monitors/
{{% /tab %}}
{{< /tabs >}}

> ## Migrate locked monitors to use RBAC

Author: ```suggestion
## Migrate monitors from locked to restricted roles
```

For more information on how to update your monitors' definitions, see the [Edit a monitor API endpoint documentation][3] and the dedicated [Monitor API Options guide][4].

### UI

All new monitors created from the UI use the `restricted_roles` parameter. 
All monitors also display the role restriction option regardless of the underlying mechanism:

{{< img src="/monitors/guide/monitor_rbac_non_restricted.jpg" alt="RBAC Non Restricted Monitor"  >}}

Datadog updates existing monitor definitions from the old locked mechanism to the new role restriction one whenever a monitor is saved.

Below are some instructions to help you determine how to proceed in case you need to save a monitor that is using the locked mechanism.

#### Locked monitors (`locked:true`) edited by creator or user with Datadog Admin Role

You are a user with [Datadog Admin Role][2] and/or are the creator of the monitor. You are editing a locked monitor and seeing the below warning:

```

> This monitor is using the `locked` attribute: only its creator and admins can edit it. `locked` is deprecated in favor of `restricted_roles`. On save, the monitor will be automatically updated to use a `restricted_roles` attribute set to all roles with Admin permissions. 

Author: ```suggestion
This monitor is using the locked attribute: only its creator and admins can edit it. locked is deprecated in favor of restricted_roles. On save, the monitor will be automatically updated to use a restricted_roles attribute set to all roles with Admin permissions. 
```

> If there is no specific change you want to apply to this monitor’s permissions, click Save. If you want to update this monitor’s permissions, read [this doc](https://docs.datadoghq.com/monitors/guide/how-to-set-up-rbac-for-monitors/?tab=ui#locked-monitors-lockedtrue-edited-by-creator-or-user-with-datadog-admin-role).

Author: ```suggestion
If there is no specific change you want to apply to this monitor’s permissions, click Save. If you want to update this monitor’s permissions, read this doc.
```

**1. You do not want to change anything about your monitor permissions**

Save the monitor. Datadog automatically migrates the monitor from the locked mechanism to restricted roles. Any other updates you made to your monitor, such as threshold update or message, are saved at the same time.

Clicking **Save** without making any changes also performs the update for your monitor.

**2. You want to allow all users to edit this monitor**

Save the monitor, causing Datadog to migrate it to restricted roles. Reopen the editing page. In the **Restrict editing of this monitor to** dropdown menu, remove all roles. Click **Save** a second time.

**3. You want your monitor to be restricted to some roles, but not the exact equivalent of the previous locked option**

Select the roles you want this monitor to be restricted to. This will override the locked equivalent set of roles to only the ones you specify.
  
#### Locked monitors (`locked:true`) edited by non creator or user without Datadog Admin Role

You are a user without the [Datadog Admin Role][2] and are not the creator of the monitor. You edit a locked monitor and see the warning below:

```

> This monitor is locked: only its creator and admins can edit it. Read more [here](https://docs.datadoghq.com/monitors/guide/how-to-set-up-rbac-for-monitors/?tab=ui#locked-monitors-lockedtrue-edited-by-non-creator-or-user-without-datadog-admin-role).

Author: ```suggestion
This monitor is locked: only its creator and admins can edit it. Read more here.
```

```

This monitor is locked. Reach out to a user with the [Datadog Admin Role][2] or to the creator of the monitor and ask them to add one of your roles to the monitor role restrictions. Your admin will have to follow steps two or three above for [locked monitors](#locked-monitors-lockedtrue-edited-by-creator-or-user-with-datadog-admin-role).

> **Note:** The discrepancy you are seeing between the warning and the option is expected. The warning reflects the current state of the monitor that is still using the locked parameter. The option reflects the role restriction option your monitor will be updated to once a user with the [Datadog Admin Role][2] or the monitor's creator edits and saves it. Once saved, the warning will disappear and the adequate restricted roles will populate the dropdown.

Author: ```suggestion
**Note:** The discrepancy you are seeing between the warning and the option is expected. The warning reflects the current state of the monitor that is using the locked parameter. The option reflects the role restriction option your monitor will be updated to once a user with the [Datadog Admin Role][2] or the monitor's creator edits and saves it. Once the monitor is saved, the warning disappears and the appropriate restricted roles populates the dropdown.
```

> **Note:** The discrepancy you are seeing between the warning and the option is expected. The warning reflects the current state of the monitor that is still using the locked parameter. The option reflects the role restriction option your monitor will be updated to once a user with the [Datadog Admin Role][2] or the monitor's creator edits and saves it. Once saved, the warning will disappear and the adequate restricted roles will populate the dropdown.

Author: ```suggestion
**Note:** The discrepancy you see between the warning and the option is expected. The warning reflects the current state of the monitor that is still using the locked parameter. The option reflects the role restriction option your monitor will be updated to once a user with the [Datadog Admin Role][2] or the monitor's creator edits and saves it. Once saved, the warning will disappear and the adequate restricted roles will populate the dropdown.
```

Author: Done

> **Note:** The discrepancy you are seeing between the warning and the option is expected. The warning reflects the current state of the monitor that is still using the old locked parameter. The option reflects the new role restriction option your monitor will be updated to once a user with a [Datadog Admin Role][2]/the monitor's creator edits and saves it. Once saved, the warning will disappear and the adequate restricted roles will populate the dropdown.

Author: ```suggestion
**Note:** The discrepancy you are seeing between the warning and the option is expected. The warning reflects the current state of the monitor that is still using the `locked` parameter. The option reflects the role restriction option your monitor will be updated to once a user with the [Datadog Admin Role][2]/the monitor's creator edits and saves it. Once saved, the warning will disappear and the adequate restricted roles will populate the dropdown.
```

> **Note:** The discrepancy you are seeing between the warning and the option is expected. The warning reflects the current state of the monitor that is using the locked parameter. The option reflects the role restriction option your monitor will be updated to once a user with the [Datadog Admin Role][2] or the monitor's creator edits and saves it. Once the monitor is saved, the warning disappears and the appropriate restricted roles populate the dropdown.

Author: ```suggestion
**Note:** The discrepancy you see between the warning and the option is expected. The warning reflects the current state of the monitor that is using the locked parameter. The option reflects the role restriction option your monitor will be updated to once a user with the [Datadog Admin Role][2] or the monitor's creator edits and saves it. Once the monitor is saved, the warning disappears and the appropriate restricted roles populate the dropdown.
```
